spoc/fragments.py

```diff
@@ -79,41 +83,58 @@ class FragmentExpander:
     """Expands n-way fragments over sequencing reads
     to yield contacts."""
 
-    def __init__(self, number_fragments: int) -> None:
+    def __init__(self, number_fragments: int, contains_metadata: bool = True) -> None:
         self._number_fragments = number_fragments
+        self._contains_metadata = contains_metadata
+        self._schema = ContactSchema(number_fragments, contains_metadata)
 
     @staticmethod
-    def _add_suffix(row, suffix, is_labelled):
+    def _add_suffix(row, suffix:int, contains_metadata:bool) -> Dict:
         """expands contact fields"""
         output = {}
-        for key in ContactSchema.get_contact_fields(is_labelled):
+        for key in ContactSchema.get_contact_fields(contains_metadata):
             output[key + f"_{suffix}"] = getattr(row, key)
         return output
 
+    def _get_expansion_output_structure(self) -> pd.DataFrame:
+        """returns expansion output dataframe structure for dask"""
+        return pd.DataFrame(columns=list(self._schema._schema.columns.keys()) + ['level_2']).set_index(["read_name", 'read_length', 'level_2'])
+
+    def _expand_single_read(self, read_df: pd.DataFrame, contains_metadata:bool) -> pd.DataFrame:
+        """Expands a single read"""
+        if len(read_df) < self._number_fragments:
+            return pd.DataFrame()
+
+        rows = list(
+            read_df.sort_values(["read_start"], ascending=True)
+            .assign(pos_on_read=lambda x: np.arange(len(x)))
+            .itertuples()
+        )
+
+        result = []
+        for alignments in combinations(rows, self._number_fragments):
```

Comment:
Yes, I would say so. If we want to implement that, I would recommend refactoring this into a separate generator method, since the logic will most likely be a little more complex.